datalad/distribution/publish.py

```diff
@@ -426,8 +429,18 @@ def _publish_dataset(ds, remote, refspec, paths, annex_copy_options, force=False
         # even if we already fetched above we need to do it again
         if is_annex_repo:
             lgr.debug("Obtain remote annex info from '%s'", remote)
-            ds.repo.fetch(remote=remote)
-            ds.repo.merge_annex(remote)
+            try:
+                ds.repo.fetch(remote=remote)
+                ds.repo.merge_annex(remote)
+            except CommandError as exc:
+                # We might be publishing for the first time to this
+                # possibly special (e.g. LFS) annex special remote,
+                # and no HEAD was known
+                if ("Couldn't find remote ref HEAD" in exc.stderr) and \
+                   ('%s/HEAD' % remote not in ds.repo.get_remote_branches()):
```

Comment:
Failing to find HEAD on the remote and the existence of the local `refs/remotes/<remote>/HEAD` are distinct things.  If a user doesn't like recording a default branch for the remote, they can `git remote set-head <remote> --delete`, and it won't be created again.

While it may seem unlikely that the remote was successfully fetched from, then locally the remote's HEAD was removed, and then for a later fetch the remote HEAD couldn't be found, I don't see a benefit of making this assumption.  Would just checking whether any refs are under `refs/remotes/<remote>` be a reliable enough indicator that we shouldn't raise the error in this case?  Something like the (untested)

```diff
diff --git a/datalad/distribution/publish.py b/datalad/distribution/publish.py
index ed2ad6cec8..5aa6250790 100644
--- a/datalad/distribution/publish.py
+++ b/datalad/distribution/publish.py
@@ -436,10 +436,7 @@ def _publish_dataset(ds, remote, refspec, paths, annex_copy_options, force=False
                 # We might be publishing for the first time to this
                 # possibly special (e.g. LFS) annex special remote,
                 # and no HEAD was known
-                if ("Couldn't find remote ref HEAD" in exc.stderr) and \
-                   ('%s/HEAD' % remote not in ds.repo.get_remote_branches()):
-                    pass
-                else:
+                if next(ds.repo.for_each_ref_(pattern="refs/remotes/" + remote):
                     raise
 
         # Note: git's push.default is 'matching', which doesn't work for first
```

---

As a tangent, I wonder if a non-zero exit code for an empty repository is specific to the hosting service or git version.  With git 2.25.1, if I add an empty gitlab repo or regular ssh target as a remote and fetch, the fetch call doesn't fail; it just doesn't bring anything down.

Reply:
sounds reasonable, I will add the change, but will delegate to `get_remote_branches` use of `refs/remotes/` ;-)
FWIW -- system git I have is 2.25, while the one bundled with git annex is (and probably was) 2.24.  So, in your case if you run the snippet in #4199 (sorry I did not make it totally cut/pasteable) -- it works on master (or ideally 0.12.2) version?

Reply:
>  but will delegate to `get_remote_branches` use of `refs/remotes/` ;-)

Okay.  The advantage of using `refs/remotes/<remote>` is that we have potentially less output to process, but, given that we're working within an exception that should not be triggered often, it's unlikely that it matters.

Reply:
>  So, in your case if you run the snippet in #4199 (sorry I did not make it totally cut/pasteable) -- it works on master (or ideally 0.12.2) version?

Sorry, I haven't taken the time to put datalad into the mix.  I'm just talking about when I create an empty repo on gitlab or an plain ssh server, add that as a remote, and `git fetch` it.  I tried with an empty repo on github, and `git fetch` (via ssh and https) exits with status of 0.

Based on the description of the problem, I think that's the state things are in when this call to `fetch` in `publish` fails, but perhaps it's some other in between state, and that is the reason for the difference.

Reply:
Stepping through fetch.c, the essential piece to trigger "fatal: Couldn't find remote ref HEAD" is that the remote does not have `remote.*.fetch` set.

Reply:
Ha. So may be we shouldn't even fetch for such remotes?

Reply:
> So may be we shouldn't even fetch for such remotes?

Given both `clone` and `remote add` will populate `.fetch` values, I think that'd be a reasonable approach, though I don't mind sticking with the current one either.

Reply:
> So may be we shouldn't even fetch for such remotes?

I've updated the PR to go with this approach.

<details>
<summary>range-diff</summary>

```
1:  19854724e1 = 1:  ac1871c47e BF: allow for fetch to fail due to HEAD not yet been available on a remote for annex
-:  ---------- > 2:  676d360692 BF: publish: Don't fetch if remote.*.fetch is not set
```

</details>